README.md

```diff
@@ -112,52 +136,111 @@ func main() {
+```
+
+The following environment variables help determine authentication configuration:
+
+- `AZURE_ENVIRONMENT`: Specifies the Azure Environment to use. If not set, it
+  defaults to `AzurePublicCloud`. Not applicable to authentication with Managed
+  Service Identity (MSI).
+- `AZURE_AD_RESOURCE`: Specifies the AAD resource ID to use. If not set, it
+  defaults to `ResourceManagerEndpoint` for operations with Azure Resource
+  Manager.
+
+## Authentication Details
+
+The previous is the first and most recommended of several authentication
+options offered by the SDK because it allows seamless use of both service
+principals and [Azure Managed Service Identity][]. Other options are listed
+below.
+
+> Note: If you need to create a new service principal, run `az ad sp
+> create-for-rbac -n "<app_name>"` in the
+> [azure-cli](https://github.com/Azure/azure-cli). See [these
+> docs](https://docs.microsoft.com/cli/azure/create-an-azure-service-principal-azure-cli?view=azure-cli-latest)
+> for more info. Copy the new principal's ID, secret, and tenant ID for use in
+> your app, or consider the `--sdk-auth` parameter for serialized output.
+
+[Azure Managed Service Identity]: https://docs.microsoft.com/en-us/azure/active-directory/msi-overview
+
+* The `auth.NewAuthorizerFromEnvironment()` described above creates an authorizer
+from the first available of the following configuration:
+
+    1. **Client Credentials**: Azure AD Application ID and Secret.
+
+        - `AZURE_TENANT_ID`: Specifies the Tenant to which to authenticate.
+        - `AZURE_CLIENT_ID`: Specifies the app client ID to use.
+        - `AZURE_CLIENT_SECRET`: Specifies the app secret to use.
+
+    2. **Client Certificate**: Azure AD Application ID and X.509 Certificate.
+
+        - `AZURE_TENANT_ID`: Specifies the Tenant to which to authenticate.
+        - `AZURE_CLIENT_ID`: Specifies the app client ID to use.
+        - `AZURE_CERTIFICATE_PATH`: Specifies the certificate Path to use.
+        - `AZURE_CERTIFICATE_PASSWORD`: Specifies the certificate password to use.
+
+    3. **Resource Owner**: Azure AD User and Password. This grant type is *not
+       recommended*, use client credentials (service principals) instead.
+
+        - `AZURE_TENANT_ID`: Specifies the Tenant to which to authenticate.
+        - `AZURE_CLIENT_ID`: Specifies the app client ID to use.
+        - `AZURE_USERNAME`: Specifies the username to use.
+        - `AZURE_PASSWORD`: Specifies the password to use.
+
+    4. **Azure Managed Service Identity**: Delegate credential management to the
+       platform. Requires that code is running in Azure, e.g. on a VM. All
+       configuration is handled by Azure. See [Azure Managed Service
+       Identity](https://docs.microsoft.com/en-us/azure/active-directory/msi-overview)
+       for more details.
+
+* The `auth.NewAuthorizerFromFile()` method creates an authorizer using tokens
+  serialized and cached by the [Azure CLI][]. Follow these steps to utilize:
```

Comment:
I believe adding this method was a mistake. It means that we've defacto taken a dependency on the CLI without actually declaring it or enforcing it in any way. For the sake of the README, I'd try to stay vague about the source of these files. Honestly, I may just skip it for now, and look into whether or not we have a means of generalizing the source of files with the schema.

For getting at the tokens as generated by the CLI, I'd rather take a true dependency on the CLI, and shell out with the command `az account get-access-token`.

Reply:
First of all the description seems wrong. not sure what you mean that it uses tokens serialized and cached by the Azure CLI. All this method does is taking doing client credentials auth but with the info in a file instead of env var

---
In reply to: [195178171](https://github.com/Azure/azure-sdk-for-go/pull/1911#discussion_r195178171) [](ancestors = 195178171)

Reply:
@marstr@microsoft.com  if that was your concern its not an issue. That being said, im not a big fan of the auth file thing since its format originated from CLI not from AAD and then it seems all the SDK's kind of went along with it.

---
In reply to: [195193861](https://github.com/Azure/azure-sdk-for-go/pull/1911#discussion_r195193861) [](ancestors = 195193861,195178171)

Reply:
so maybe we can remove it but maybe for some customers its useful. so maybe remove it form the readme or add an experimental or some kind of info showing that is not something 100% standard from what I am aware at least.

---
In reply to: [195194277](https://github.com/Azure/azure-sdk-for-go/pull/1911#discussion_r195194277) [](ancestors = 195194277,195193861,195178171)

Reply:
@marstr this isn't using `~/.azure/accessTokens.json`, the CLI's internal serialization file. It's using the format generated by `az ad sp create-for-rbac --name 'delete-me' --sdk-auth`. This is supported in other SDKs and we should continue supporting it here at least for now.

@vladbarosan good point that it would be better if this came from AAD, but we've decided to define and pursue this format in the SDKs/CLIs team now, in fact we're working on standards to apply across all languages.